tools/release

```diff
@@ -0,0 +1,95 @@
+#
+# It takes two mandatory arguments:
+#
+#   - the URL to clone the repository from,
+#     eg. 'https://github.com/cockpit-project/cockpit'.
+#
+#   - the version to release, which is also the tag name, eg. '215'.
+#
+# The release tarball contains a number of files that are not directly in
+# version control:
+#   - the usual automake stuff
+#   - several submodules
+#   - pre-compiled HTML and JS code
+#
+# This script shallow-clones the specified version of Cockpit from the
+# specified URL, and downloads the version of the tasks container specified in
+# that version. It then does an offline build of the release tarball, which is
+# deterministic.
+#
+# You can use this script to verify the integrity of a particular Cockpit
+# release tarball, or even use it to create the tarball for yourself, as an
+# alternative to downloading it.
+#
+# The result of running `./autogen.sh && make dist VERSION=...` on any system
+# ought to be the same, so long as the same versions of the autotools are
+# present.  Everything else present in the tarball is definitively pinned down
+# in one way or another by the contents of the git repository.  This script is
+# a bit overkill, but it's written in hopes that someone reading it can
+# reasonably convince themselves about the origin of absolutely everything in a
+# Cockpit source release.
+#
+# If this script produces a tarball with a different checksum than the
+# officially-released one, then please report a bug.
+
+URL="$1"
+VERSION="$2"
+
+# We collect the sources into a temporary directory, then use tar to pipe them
+# into a container with no network or filesystem access, where the actual build
+# takes place. The container pipes the built source release back to us.
+
+SOURCE="$(mktemp -dt 'cockpit-build-XXXXXX')"
+trap 'rm -rf "${SOURCE}"' EXIT
+
+# Clone the release and selected submodules into the temporary directory.
+git clone \
+    --depth=1 \
+    --recurse-submodules=node_modules \
+    --recurse-submodules=vendor \
+    -b "${VERSION}" \
+    "${URL}" \
+    "${SOURCE}"
+
+# Show exactly what we're building.
+git -C "${SOURCE}" show --no-patch "${VERSION}"
+
+# Download the tasks container image used for this release.
+IMAGE="$(cat "${SOURCE}"/.cockpit-ci/container)"
+podman pull "${IMAGE}"
+
+# Build the checked out sources into the release tarball, offline.
+tar -C "${SOURCE}" -c . | \
+    podman \
+        run \
+            --rm \
+            --pull=never \
+            --network=none \
+            --log-driver=none \
+            --interactive \
+            --env=VERSION="${VERSION}" \
+            "${IMAGE}" \
+            sh -euxc '
+                (
+                    mkdir work
+                    cd work
+                    rpm -q autoconf automake
+                    automake --version
+                    autoconf --version
+                    tar x
+                    ./autogen.sh
+                    make dist VERSION="${VERSION}"
```

Comment:
We do something completely different here than what we do in `tools/make-dist` -- the script which we use absolutely everywhere else: Our CI, packit, local testing, and even our flatpak.

It would be nice to put this shell script into tools/make-dist to unify these, and declare the "can run in configured tree" use case obsolete (does anyone care?), and use the same script for everything.

Reply:
I'd say that we're doing two separate things, with different intents.

The `make-dist` script builds a release out of the current directory (with any changes that might be lying around in the tree), with a `+git` version number, no compression, and doesn't use a container to do it.

I also wanted to make use of the `make-dist` script from the outside script.  I suppose we could pass the version and the `XZ_OPT` override....

I think another reason I avoided doing that is that it's one more thing for the person reading the outer script to try to understand...

Let me give it a try and see how it looks...

Reply:
I mean, just looking at it now, big meh.  The only two lines that we could reasonably replace there are

```sh
./autogen.sh
make dist VERSION="${VERSION}"
```

It's just not worth it, imho.

Reply:
From the other side around, trying to use the container approach to build the current working directory of cockpit, we'd basically end up having to reinvent something like the old unit-tests start script (to deal with changes in the worktree, caching submodules, git worktrees, selinux, etc., etc.).

I think it's pretty unfortunate, but these are two very separate things.

One thing we might want to do is to try and increase our use of the `release` script introduced here for more things that happen off of GitHub... although to be honest, nothing really comes to mind...

Reply:
Ah I see, the `tar` gets in between. 

I didn't actually mean to invoke the container during test/image-prepare -- that would not work anyway, and also a bit stupid as this *already* runs in the tasks container. I really just meant the inner shell script, not the podman invocation around it. That's why I usually prefer volumes (the source could be mounted as read-only into the container even).

My main worry is that this code path is completely untested until the release, so we could be producing broken release tarballs and never notice. Things could randomly appear on stdout, or there's a bug in that script, etc.

Reply:
I agree.